### PR TITLE
Refactor test card

### PR DESCRIPTION
**Pull request by mpbrown**
# FRONTEND PULL REQUEST

## Related Issue

- #6079 

## Changes Proposed

- Adds a new `TestCard` component used instead of the existing `QueueItem` if the `testCardRefactorEnabled` feature flag is active.
- `TestCard` uses a collapisble design which will display the `TestCardForm` as a child component when the card is expanded. TestCardForm is where the majority of the conduct tests business logic is used while TestCard primarily serves as the header and overall form container.
- Currently the test cards are expanded by default to make it an easier transition for users unfamiliar with the new collapsible design, but this should be updated in the future.

### AOE
- Test questionnaire modal will no longer appear when starting a patient test from either the search bar or the patients page.
- AOE questions will now be shown directly on the `TestCardForm`. The COVID AOE questions are used by default, but other disease specific AOE form components can be swapped when additional supported diseases are added.
- Test result delivery preferences will no longer be included when saving AOE responses. The backend defaults to using whatever preferences the patient already has saved on their profile.

### Test Results
- The input groups for marking test results no longer use their own internal submit button.
- The radio button options for results will now show as asterisk if it is required, such as when either some or all multiplex results are reburied for submission.
- Error messages for missing test results are shown after the user attempts to submit the form. The submit button is no longer disabled for this reason.

### Timer
- Test timer design has been updated which will be present on both the new `TestCard` and existing `QueueItem`.

## Additional Information
- TestCardForm uses `useReducer` to manage the form state instead of multiple `useState` hooks. This was chosen because some state properties depend on each other or on the previous state and encapsulating this state management in a reducer with discrete dispatch actions improves the testability and readability of the main TestCardForm.
- `TestCardForm.utils` contains some custom hooks and helper functions with the goal of balancing just enough abstraction to improve readability and understanding of how the test card works overall.

## Testing

- Test out the new changes on [dev7](https://dev7.simplereport.gov/app/admin?facility=f5078710-8218-4308-861a-08dee77343ac).
- Please specifically test out different states of the test card and try making fast changes or unexpected inputs to test the behavior of the form saving live changes and receiving any updates from the server. 
    - Keep in mind that if someone else is testing on the same patient at the same time that may lead to unexpected but intentional behavior where an update from the server might overwrite a change you previously made.
- Unit test suite for the new `TestCard` and `TestCardForm` is still in progress.

## Screenshots / Demos

Multiplex with all results
image

Multiplex missing results
image

Flu only device and custom backdate
image

Incomplete AOE questions confirmation modal before submitting
image

Invalid test date
image

Collapsed view
image

### CODE REVIEW

**Comment by mpbrown** [comment]
FYI on `dev7` I'm experiencing an unexpected 403 forbidden error when I try answering an AOE question, but this might just be related to needing to update my user permissions for my logged in organization on dev7 since answering and saving AOE works on local deployment.

File: frontend/src/app/testQueue/TestCard/TestCard.stories.tsx
Comment (johanna-skylight) [inline frontend/src/app/testQueue/TestCard/TestCard.stories.tsx:14]: Does this unused variable warning need to be fixed?

File: frontend/src/app/testQueue/TestCard/TestCard.test.tsx
Comment (johanna-skylight) [inline frontend/src/app/testQueue/TestCard/TestCard.test.tsx:683]: Does this commented code need to be removed?

File: frontend/src/app/testQueue/TestCard/TestCard.test.tsx
Comment (johanna-skylight) [inline frontend/src/app/testQueue/TestCard/TestCard.test.tsx:840]: these type methods probably need an `await`

File: frontend/src/app/testQueue/TestCard/TestCard.tsx
Comment (johanna-skylight) [inline frontend/src/app/testQueue/TestCard/TestCard.tsx:88]: Could this modal be moved to its own file?

File: frontend/src/app/testQueue/TestCardForm/TestCardForm.tsx
Comment (johanna-skylight) [inline frontend/src/app/testQueue/TestCardForm/TestCardForm.tsx:348]: Can this modal be moved to its own file as well?

File: frontend/src/app/testQueue/addToQueue/AddToQueueSearch.tsx
Comment (johanna-skylight) [inline frontend/src/app/testQueue/addToQueue/AddToQueueSearch.tsx:101]: Comment probably needs to be removed

**Comment by johanna-skylight** [comment]
Did some testing in dev7 and saw below:
- Tooltip for mobile view is getting cut.
Screenshot 2023-10-13 at 2 21 24 PM

- I loaded a COVID device and didn't see AoE form but then the form was displayed when picking other COVID device.
Screenshot 2023-10-13 at 2 22 10 PM
Screenshot 2023-10-13 at 2 24 40 PM

- When trying to click on answer in the AoE form I kept getting backend errors.
Screenshot 2023-10-13 at 2 24 52 PM

- Even though the AoE answers were not able to be saved I was able to submit the test result and the warning about the AoE form not being filled was not displayed.

**Review by johanna-skylight — COMMENTED** [review]
Thanks for working on this!. I have left some comments.

File: frontend/src/app/testQueue/TestCard/TestCard.test.tsx
Comment (mpbrown) [reply frontend/src/app/testQueue/TestCard/TestCard.test.tsx:827]: Turns out that `getByTestId` is synchronous so eslint warns against adding await here

**Comment by mpbrown** [comment]
> * I loaded a COVID device and didn't see AoE form but then the form was displayed when picking other COVID device.

@johanna-skylight it looks like on dev7 that QIAstat device doesn't have any explicit supported diseases configured. Should we default to handling it as a COVID only device in that case or should this device be considered invalid?

Screenshot 2023-10-13 150233

> * Even though the AoE answers were not able to be saved I was able to submit the test result and the warning about the AoE form not being filled was not displayed.

Hmm I also ran into that 403 error so I'll take a closer look at dev7 permissions since so far I can't replicate it on a local deployment

File: frontend/src/app/testQueue/TestCard/TestCard.test.tsx
Comment (mpbrown) [reply frontend/src/app/testQueue/TestCard/TestCard.test.tsx]: Updated to use await with `findByTestId` instead, thanks for spotting this!

**Comment by mpbrown** [comment]
Recent updates have fixed the 403 error for saving AOE responses and fixed the issue of no AOE questions showing when the selected device doesn't have any explicit supported tests (seen in legacy covid only devices in some of the dev environments)

Design is also taking a look at the mobile layout including how we handle tooltip info, but that will likely be in another story since this PR doesn't enable the feature flag for prod

File: backend/src/main/resources/application-azure-demo.yaml
Comment (DanielSass) [inline backend/src/main/resources/application-azure-demo.yaml:16]: Do we need this snippet for the other envs?
Or a default set in the base application yaml?

Comment (mpbrown) [reply backend/src/main/resources/application-azure-demo.yaml:16]: A default was set in the separate feature flag PR #6652 that had it enabled except for being disabled on prod. These changes also disable it on the other higher level environments except for `training` which design is planning to use for user testing sessions with the new card

**Comment by nathancrtr** [comment]
I didn't see the ticket or design specify whether the test cards should default to the collapsed state or the expanded one. It looks like they're expanded by default? Just wanting to call this out to confirm 👍

File: frontend/src/app/testQueue/TestCardForm/TestCardForm.tsx
Comment (DanielSass) [inline frontend/src/app/testQueue/TestCardForm/TestCardForm.tsx:140]: I know I've had to disable this linting error before, but with how consistently we're ignoring it - is there a more correct way of doing this?  Or can we update the linter to not complain about this if its not actually an issue?

**Comment by nathancrtr** [comment]
It doesn't feel important to correct this under this particular ticket, but just calling out a little wonky layout behavior:

Screen Shot 2023-10-20 at 2 21 35 PM

This is in the case where a facility has its "default" device removed while a test card is open.

**Review by DanielSass — APPROVED** [review]

**Comment by zdeveloper** [comment]
> It doesn't feel important to correct this under this particular ticket, but just calling out a little wonky layout behavior:
> 
> Screen Shot 2023-10-20 at 2 21 35 PM
> This is in the case where a facility has its "default" device removed while a test card is open. And ditto, I guess, for doing the same with a device's default specimen type:
> 
> Screen Shot 2023-10-20 at 2 24 08 PM
> Again, really nitpicky and best looked at elsewhere, IMO, just shouting it out for now.

Yeah I agree, but this is a super edge case, especially with relying on automatic livdsync, I can see us never really deleting a device or specimen since we already corrected all of our devices info.

File: frontend/src/app/testQueue/TestQueue.tsx
Comment (emyl3) [inline frontend/src/app/testQueue/TestQueue.tsx:166]: I think we may want to add a `finally` block here to `setStartTestPatientId(null);` because we run into issues if we add a patient via the patients page to the test queue


Uploading error toast bug.mov…

File: frontend/src/app/testQueue/TestCardForm/TestCardForm.utils.tsx
Comment (emyl3) [inline frontend/src/app/testQueue/TestCardForm/TestCardForm.utils.tsx:213]: is there a reason we are surfacing the errors as an alert at the top of the page as opposed to the inline form error pattern we use elsewhere? 🤔 
Screenshot 2023-10-20 at 15 52 07

File: frontend/src/app/testQueue/TestCardForm/TestCardForm.tsx
Comment (mpbrown) [reply frontend/src/app/testQueue/TestCardForm/TestCardForm.tsx:140]: Thanks for mentioning this! After I looked into it a bit more, the React team [strongly recommends keeping the rule](https://github.com/facebook/create-react-app/issues/6880#issuecomment-485912528) and reworking the code to handle it properly. I'm experimenting with the code and we may be able to remove some of these useEffects entirely, so I'll create a follow up PR for this.

Some of this effect code will benefit significantly when React officially releases the [new `useEffectEvent` hook](https://react.dev/learn/separating-events-from-effects#declaring-an-effect-event) on a stable version. For example, we need to control for race conditions between the frontend saving edits and the periodic live sync with any updates from the backend. 

When we get new values from the backend, we have to check whether the mutations are still loading and _although those loading booleans are reactive values_ that React says should be included in the dependency array, we don't want those in the array because then the effect will be triggered whenever those change and potentially cause sync issues.
```
useEffect(() => {
    // don't update if there are unsaved dirty changes or if still awaiting saved edits
    if (state.dirty || editQueueItemMutationLoading || updateAoeMutationLoading)
      return;
    dispatch({
      type: TestFormActionCase.UPDATE_WITH_CHANGES_FROM_SERVER,
      payload: testOrder,
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [testOrder]);
```

So once `useEffectEvent` is officially added, we'll be able to separate reactive values that are actual effect dependencies from ones that are only used for conditional logic.

File: frontend/src/app/testQueue/TestQueue.tsx
Comment (mpbrown) [reply frontend/src/app/testQueue/TestQueue.tsx:166]: Great bug find! I'll push an update though it may have to be added in the `catch` block to only reset when there's an error since that id is still used after the mutation succeeds

File: frontend/src/app/testQueue/TestCardForm/TestCardForm.utils.tsx
Comment (mpbrown) [reply frontend/src/app/testQueue/TestCardForm/TestCardForm.utils.tsx:213]: I'll defer to design on the specific reason but I believe this was related to handling when we have multiple radio groups for a multiplex device. Before it gets enabled in prod, we'll be updating the test card design based on upcoming user feedback so we could address it in that future story

File: frontend/src/app/testQueue/TestCardForm/TestCardForm.tsx
Comment (emyl3) [inline frontend/src/app/testQueue/TestCardForm/TestCardForm.tsx:460]: just a heads up in the original queue card, G&E implemented a check to make sure we don't show RSV devices https://github.com/CDCgov/prime-simplereport/blob/ed7620b94596269419adc6cc8e1b7bcd4fe7a7b8/frontend/src/app/testQueue/QueueItem.tsx#L762 we would want that here as well

**Review by DanielSass — APPROVED** [review]

**Review by emyl3 — CHANGES_REQUESTED** [review]
Nice work, Mike! Left my first bunch of comments. 

Do you have access to Deque's axe dev tool? I think it will be helpful to run particularly against multiple test cards in the conduct test page.

File: frontend/src/app/testQueue/TestTimer.tsx
Comment (mpbrown) [reply frontend/src/app/testQueue/TestTimer.tsx:243]: I agree this should be added on the aria label, although I think we should create a separate ticket for it since it'll also affect the existing queue item component which will still be used in prod until the new test card feature flag is enabled

**Comment by mpbrown** [comment]
@emyl3 thanks for the reminder about the axe devtools extension! I pushed some updates that improve accessibility. Once the PR is approved, the design team will be conducting usability testing using the demo environment. The new test card feature flag will remain disabled in the prod environment

**Review by DanielSass — APPROVED** [review]

File: frontend/src/app/testQueue/TestCardForm/diseaseSpecificComponents/MultiplexResultInputGroup.tsx
Comment (emyl3) [inline frontend/src/app/testQueue/TestCardForm/diseaseSpecificComponents/MultiplexResultInputGroup.tsx:288]: Screenshot 2023-10-23 at 15 26 27
Something we can also flag for design is marking the labels as required when we select "Mark test as inconclusive" is misleading

Comment (DanielSass) [reply frontend/src/app/testQueue/TestCardForm/diseaseSpecificComponents/MultiplexResultInputGroup.tsx:288]: that's a good catch 🔍

File: frontend/src/app/testQueue/TestCard/TestCard.tsx
Comment (emyl3) [inline frontend/src/app/testQueue/TestCard/TestCard.tsx:90]: Screenshot 2023-10-23 at 15 20 24

Looks like the `<Card>` element is a `<li>` and needs to be nested under `<ul>`

**Review by emyl3 — APPROVED** [review]
flagged a couple more things but am approving since it sounds like we'll be doing more work on this later and we need to get this out for design/research soon! functionality-wise looks good

File: frontend/src/app/testQueue/TestCardForm/diseaseSpecificComponents/MultiplexResultInputGroup.tsx
Comment (mpbrown) [reply frontend/src/app/testQueue/TestCardForm/diseaseSpecificComponents/MultiplexResultInputGroup.tsx:288]: We're likely going to remove that inconclusive checkbox altogether in favor of simplifying the result validation rules for multiplex devices. That change should be in the upcoming PR for supporting HIV on the test card

File: frontend/src/app/testQueue/TestCard/TestCard.tsx
Comment (mpbrown) [reply frontend/src/app/testQueue/TestCard/TestCard.tsx:90]: Noticed this on the axe scan, but holding off on changing it until the follow-up PR because a few issues popped up with nesting the queue under a single `ul` so additional refactor and updating some tests would be needed

**Pull request merged**
